### ISSUE TITLE
BatchNorm with statistics from whole dataset

### ISSUE THREAD

**Issue by RaffEdwardBAH** [fchollet/keras#3949]
I've had a number of issues recently with the batch-norm statistics being "dirty" for some very large networks we are training. It seems the exponential average used has too much history from earlier iterations and the weights have drifted so far away that they are no longer useful. We see this issue manifest itself as a huge difference in training accuracy / loss, and predictiction accuracy /loss _on the training data_. While this issue "resolves" itself if we let the network continue to train for several more epochs, these models are huge and time consuming - it would be faster to just do an epoch of predictions and get the correct statistics as used in the original paper. 

Is there a way to get keras to do this? I'm not seeing an obvious place to try and modify the keras code to add this functionality.

**Comment by robertomest** [comment]
This is not the solution you are looking for, but have you tried a smaller momentum value? Keras' standard value is quite high (0.99). Torch, if I recall correctly, uses 0.9. Maybe that would solve your issue without the trouble of coding a different solution.

**Comment by RaffEdwardBAH** [comment]
I have not yet, no. I want to avoid that due to the immense training time of our models. For us, it would be far more practical if we could load in an already trained model and get the whole data set statistics. 

Did the default value change? I thought it used 0.9 originally, but checking the code -- it indeed uses 0.99 right now.

**Comment by robertomest** [comment]
It seems to have been changed on this commit https://github.com/fchollet/keras/commit/3533912016c64f1f03516657c1d93f02a11e5c67. I don't know of an easy way of using the whole dataset statistics though.

**Comment by tzachar** [comment]
I have raised this issue a while back, but got no response (see #3571).
The best I could come up with was a bit hacky:
1) first, change the momentum variable of bach norm to be a tensor.
2) post train, go over all layers of the model besides the bn layers, marking  them non trainable
3) perform a final training epoch, while adjusting the momentum of all bn layers accordingly: first batch, momentum should be 0. second batch, 0.5., etc. 

Don't know if its worth it. I ended up decreasing the momentum below 0.9 to get good results.
A better approach would be to change the bn layer to include a post train phase, where the statistics are summed and averaged at the end. This probably entails adding a post-train phase to keras models.

Cheers.
